### ISSUE TITLE
To launch embedded Tomcat, 1.2.x requires the temp directory to exist whereas 1.1.x did not

### ISSUE THREAD

**Issue by henrybuckley** [spring-projects/spring-boot#3307]
I'm trying out spring boot for the first time using cygwin on Windows 8.1 with the Springboot 1.2.4.RELEASE version. When I invoked `spring run` on the following 'Hello World' groovy script:

```groovy
@RestController
class App {
      @RequestMapping("/")
      def home() {
          "Hello, world!"
      }
}
```
the following stack trace appeared in the console: 

<pre>org.springframework.context.ApplicationContextException: Unable to start embedded container; nested exception is org.springframework.boot.context.embedded.EmbeddedServletContainerException: Unable to create Tomcat tempdir
        at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.onRefresh(EmbeddedWebApplicationContext.java:133)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:474)
        at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:118)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:686)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:320)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.springframework.boot.cli.app.SpringApplicationLauncher.launch(SpringApplicationLauncher.java:69)
        at org.springframework.boot.cli.command.run.SpringApplicationRunner$RunThread.run(SpringApplicationRunner.java:148)
Caused by: org.springframework.boot.context.embedded.EmbeddedServletContainerException: Unable to create Tomcat tempdir
        at org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedServletContainerFactory.createTempDir(TomcatEmbeddedServletContainerFactory.java:390)
        at org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedServletContainerFactory.getEmbeddedServletContainer(TomcatEmbeddedServletContainerFactory.java:142)
        at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.createEmbeddedServletContainer(EmbeddedWebApplicationContext.java:157)
        at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.onRefresh(EmbeddedWebApplicationContext.java:130)        at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.onRefresh(EmbeddedWebApplicationContext.java:130)
        ... 10 common frames omitted
Caused by: java.io.IOException: The system cannot find the path specified
        at java.io.WinNTFileSystem.createFileExclusively(Native Method)
        at java.io.File.createTempFile(File.java:2024)
        at java.io.File.createTempFile(File.java:2070)
        at org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedServletContainerFactory.createTempDir(TomcatEmbeddedServletContainerFactory.java:383)
        ... 13 common frames omitted
</pre>

When I switched my spring boot version to an earlier version of spring boot (e.g. 1.1.6.RELEASE) then `spring run` completed successfully. When I then switched back to the 1.2.4.RELEASE version which originally had this issue, I no longer saw the exception when running `spring run` on the groovy script, and the embedded server started successfully.

**Comment by wilkinsona** [comment]
The code in question is unchanged between 1.1.6.RELEASE and 1.2.4.RELEASE so It's not immediately clear why this should work in 1.1.6 and fail in 1.2.4. In both cases a call to  `File.createTempFile(prefix, suffix)` is made. This method creates a temporary file in the default temporary directory, the location of which is controlled by the `java.io.tmpdir` System property. I would guess that you had `java.io.tmpdir` pointing to a non-existent directory and after the failure with 1.2.4 something created that directory, so that when you tried it again it no longer failed.

In short, you should ensure that `java.io.tmpdir` points to a directory that exists and that you have the necessary permissions to write to it.

**Comment by snicoll** [comment]
I might add that `cygwin` caught me a few times with issues like this so I'd try in a regular shell as well.

**Comment by henrybuckley** [comment]
@wilkinsona You deduced correctly. I happened to be running in a Cygwin environment where the `TMP` environment variable (by default used by the Java runtime to set the `java.io.tmpdir` system property) was set to `C:\tmp` - and this directory didn't exist at that point on my system.

A suggestion for greater ease of debugging: include the directory path of the directory that failed to be created in the `EmbeddedServletContainerException` exception message.

There was one more thing I discovered, which explains why it started working for me after I reverted to the 1.1.6.RELEASE version. If I reset my `TMP` environment variable to point to a directory which doesn't exist on my system, and execute `spring run app.groovy` under the 1.1.9.RELEASE (or earlier), then the missing directory is created automatically (and the application is launched successfully). If instead I execute `spring run app.groovy` under the 1.2.0.RELEASE (or later), then the missing temporary directory is not created and the application fails to launch (as previously described).

**Comment by wilkinsona** [comment]
I've figured out why the behaviour changes between 1.1.x and 1.2.0. In 1.1.x, Boot would log to both the console and a file. In 1.2.0, that was changed to only log to the console by default. The default log file location is `${LOG_FILE:-${LOG_PATH:-${LOG_TEMP:-${java.io.tmpdir:-/tmp}}/}spring.log}`. When Logback wants to write to `spring.log` it makes the necessary parent directories. In this case, in the absence of configuration for `LOG_FILE`, `LOG_PATH`, and `LOG_TEMP`, that means it makes the directory that's pointed to by `java.io.tmpdir`. This happens before `TomcatEmbeddedServletContainerFactory` tries to create the temporary directory so it then succeeds.
